### ISSUE TITLE
Schema for .clangd config

### ISSUE THREAD

**Issue by jmigual** [clangd/clangd#1376]
Since the `.clangd` config file is a YAML file it'd be very useful to create a schema file describing the different settings and their parameters. 

This way the schema could be referenced in editors such as VS Code that take a key `$schema: <url>` and are able to autocomplete the keys in the YAML file. 

In my personal projects I've seen that even for YAML files, VS Code recognizes the JSON schema format from https://json-schema.org and is able to validate YAML files using a JSON schema.

**Comment by i-ky** [comment]
[YAML extension for VS Code](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) allows other extensions to contribute YAML schemas [via `yamlValidation` contribution point](https://github.com/redhat-developer/vscode-yaml#mapping-a-schema-in-an-extension). [clangd extension](https://github.com/clangd/vscode-clangd) can do the same.

Note that at the moment clangd provides diagnostics for configuration files via LSP. Maybe LSP can be used to provide completions for config files as well.

**Comment by hyperupcall** [comment]
I added the schema definition to SchemaStore here: https://github.com/SchemaStore/schemastore/pull/2653

Now, users of all IDEs can get the suggestions, without installing any specific extensions

**Comment by jmigual** [comment]
Then we can close this issue

**Comment by HighCommander4** [comment]
> I added the schema definition to SchemaStore here: [SchemaStore/schemastore#2653](https://github.com/SchemaStore/schemastore/pull/2653)
> 
> Now, users of all IDEs can get the suggestions, without installing any specific extensions

This is really neat, thanks!

Do you know how to set the validation up in VSCode?

**Comment by hyperupcall** [comment]
@HighCommander4 Things should Just Work :) - you may have to reload the editor or run the `JSON: Clear schema cache`, but that's it. There is [this](https://github.com/SchemaStore/schemastore/issues/2438) issue on the SchemaStore repo with some relevant discussion, but I didn't bump into that error.

On the bottom-right, there is a "No JSON Schema" button that you can click (you might have to manually set the language to YAML), and you can also choose a schema from there:

![image](https://user-images.githubusercontent.com/24364012/208328916-2634d10e-6bd0-4e06-98a0-1b8785e8c543.png)

**Comment by HighCommander4** [comment]
> I was also expecting it to issue diagnostics for unrecognized keys (e.g. if you made a typo in a key name), but I'm not getting that, not sure if that's expected.

(But clangd issues such diagnostics itself if it's running, so not a big deal.)

**Comment by HighCommander4** [comment]
One question that has occurred to me is: since the clangd config file format evolves over time (e.g. new keys are added), would it make sense to check the schema file into the clangd repo, and encourage clangd contributors to keep it up to date in the same patches that add new config functionality?

We'd still need a mechanism for keeping the copy in the SchemaStore repo up to date, but this would at least ensure that the clangd repo copy is kept up-to-date as the config format is updated, rather than someone else having to periodically circle back and track down the changes to update the schema.

**Comment by hyperupcall** [comment]
@HighCommander4 You're right, it doesn't mark typo'ed keys an error. I'm kind of indecisive about that - you're supposed to make it an error, but that would mean whenever you try to use a brand new valid key (that isn't in the schema), you get annoying red lines. Are the red lines something you would prefer in the normal circumstance?

About your other question, SchemaStore let's you reference an external schema, like to one in this repository . Also most of my PRs over there changes get accepted in a few hours, so if you do decide to keep the schema there things update prerty quick

**Comment by HighCommander4** [comment]
> @HighCommander4 You're right, it doesn't mark typo'ed keys an error. I'm kind of indecisive about that - you're supposed to make it an error, but that would mean whenever you try to use a brand new valid key (that isn't in the schema), you get annoying red lines. Are the red lines something you would prefer in the normal circumstance in that scenario?

I think such diagnostics are useful in general, but in this case since clangd itself produces them if it's running, it's probably fine for the YAML plugin to omit them to avoid duplicate diagnostics.

> About your other question, SchemaStore let's you reference an external schema, like to one in this repository.

Ah, perfect! Then I think it would make sense to keep it here. Let's reopen the issue to track that.

Are you interested in writing a patch to add it?

**Comment by hyperupcall** [comment]
> Are you interested in writing a patch to add it?

Yas! I'm just unsure where to put it? Once I do that I can do the SchemaStore PR.

**Comment by HighCommander4** [comment]
Maybe `clang-tools-extra/clangd/schema/config.json` (with `schema` being a new directory)?

**Comment by hyperupcall** [comment]
@HighCommander4 It's done - [D140462](https://reviews.llvm.org/D140462)! I added you as reviewer and I think I did everything okay? Followed contribution guide to best of my ability, but I still haven't used Phabricator or submitted patches to LLVM before
